pages/display.py

```diff
@@ -16,58 +16,38 @@ class Page:
     # Fields stored in DB. See class `pages.models.PageTemplate`
     STORED = ['name', 'h1', 'keywords', 'description', 'title', 'seo_text']
 
-    def __init__(
-        self,
-        context: typing.Dict[str, typing.Any]=None,
-        *,
-        _page: 'pages.models.Page'=None,
-        _name='',
-    ):
+    def __init__(self, page: 'pages.models.Page'=None, context: typing.Dict[str, typing.Any]=None):
         """
         Pass context at ctor, but not render method,
         because client code wants the same context for many different cases.
         """
-        self._context = context or {}
-        self._page = _page
-        self._key = _name
-
-    def __set_name__(self, instance, name):
-        # get `Page` instance to create new `Page` instance with `name`
-        page = getattr(instance, name)
-        page_with_name = Page(page._context, _page=page._page, _name=name)
-        # Set new `Page` instance with name
-        setattr(instance, name, page_with_name)
+        self._page = page
+        self.key = ''
 
     def __get__(self, instance: 'pages.models.Page', type_):
-        if instance and self._key in instance.__dict__:
-            return instance.__dict__[self._key]
         return Page(
-            {'page': instance, **self._context},
-            _page=instance,
-            _name=self._key,
+            instance,
+            {'page': instance, **instance.__dict__.get(self.key, {})}
         )
 
     def __set__(self, instance: 'pages.models.Page', value: typing.Dict[str, typing.Any]):
-        if not isinstance(value, dict):
-            raise ValueError(f'Value should be a dict')
-
-        if self._key in instance.__dict__:
-            # merge old context with new one
-            context = {**instance.__dict__[self._key]._context, **value}
-        else:
-            context = {**self._context, **value}
-
-        instance.__dict__[self._key] = Page(context, _page=instance, _name=self._key)
+        self._page = instance
+        instance.__dict__[self.key] = value
 
     def __getattr__(self, item):
         if item in self.STORED:
             return self.render(item)
         else:
             return super().__getattribute__(item)
 
+    def __set_name__(self, owner, name):
+        self.key = f'_{name}_value'
+
     def render(self, field: str):
         return (
-            self._page.template.render_field(field, context=self._context)
+            self._page.template.render_field(
```

Comment:
I suspect right now `self._page` field is shared among instances, so let's drop it and get `page` instance from `self._page.__dict__[self.key]['page']`. 

e.g.
```
class A:
    d = display.Page()
a1 = A()
a2 = A()
a1.d = {}
a2.d = {}

assert A.__dict__['d']._page == a2
a1.d.h1 # seems raise the issue, because `self._page == a2`
```

Reply:
@artemiy312 , seems display.Page is shared between models.Page instances. It's very strange for me. But  _page field is not shared, and it's strange too:
```
In [1]: from pages import display
   ...:
   ...: class A:
   ...: ^Id = display.Page()
   ...: ^I
   ...: a, b = A(), A()
   ...: print(id(a.d), id(b.d))
   ...: print(id(a.d._page), id(b.d._page))
140292117405256 140292117405256
140292117405144 140292117405200
```

Previous code had this behaviour too.

Shared page instance should bring us problems with different page templates rendering.
Created test `tests.pages.test_models.TestPage#test_display_has_unique_template` to prove templates part of the context works

Reply:
@artemiy312 , i made tests up and code takes less lines, but looks too complex and dangerous.
Global dict with both context and models.Page instance preserving now looks the most reasonable way to me.

Decide plz what of these three solutions we are going to leave

Reply:
@duker33 sorry, everything is right.
The suspicion occurred, because of `self._page = instance` in `__set__` method, that is redundant, because we pass an `instance` to a `Page` through ctor in `__get__`

So you should just drop `self._page = ...` in `__set__` 

This implementations seems to me the better one. BTW the implementation with global context looks similar except of global variable :)

Reply:
@artemiy312 , i'll remove it, sure.
About new page instance via `__get__` method. This quote from the output ahead tells us, that instance creating inside the `__get__` looks not very reliable. That's why i'm in worry about this method.

This method pretty like global var, but uses for storing instances instead. The thin moment here, that in the `render` method we can have no access to an actual `intance` (of models.Page).
The global var approach has not this disadvantage.

So, i'll merge it

```
In [1]: from pages import display
   ...:
   ...: class A:
   ...:     d = display.Page()
   ...:     
   ...: a, b = A(), A()
   ...: print(id(a.d), id(b.d))

140292117405256 140292117405256
```